spec/unit/puppet/provider/registry_key_spec.rb

```diff
@@ -26,4 +29,50 @@
       reg_key.provider.exists?.should be_false
     end
   end
+
+  describe "#purge_values" do
+    let (:guid) { SecureRandom.uuid }
+    let (:reg_path) { "#{puppet_key}\\#{subkey_name}\\Unicode-#{guid}" }
+
+    def bytes_to_utf8(bytes)
+      bytes.pack('c*').force_encoding(Encoding::UTF_8)
+    end
+
+    before(:each) do
+      # create temp registry key with Unicode values
+      Win32::Registry::HKEY_LOCAL_MACHINE.create(reg_path,
+        Win32::Registry::KEY_ALL_ACCESS) do |reg_key|
+          endash = bytes_to_utf8([0xE2, 0x80, 0x93])
+          tm = bytes_to_utf8([0xE2, 0x84, 0xA2])
+
+          # TODO: this doesn't really trigger string encoding issues
+          reg_key.write(endash, Win32::Registry::REG_SZ, tm)
+      end
+
+      # problematic Ruby codepath triggers a conversion of UTF-16LE to
+      # a local codepage which can totally break when that codepage has no
+      # conversion from the given UTF-16LE characters to local codepage
+      # a prime example is that IBM437 has no conversion from a Unicode en-dash
+      Win32::Registry.expects(:export_string).never
+
+      # also, expect that we're not using Rubys each_key / each_value which exhibit bad behavior
+      Win32::Registry.expects(:each_key).never
+      Win32::Registry.expects(:each_value).never
```

Comment:
If putting in a shim to use `Win32::Registry` under `2.0.0` and lower, but Puppet itself in `Puppet 4` / `Ruby 2.1.5` ... these need to be guarded.